src/main/java/com/arpnetworking/clusteraggregator/client/HttpSourceActor.java

```diff
@@ -0,0 +1,327 @@
+                        ));
+
+        final Optional<String> host = Optional.ofNullable(dimensionsMap.get(CombinedMetricData.HOST_KEY));
+        Optional<String> service = Optional.ofNullable(dimensionsMap.get(CombinedMetricData.SERVICE_KEY));
+        Optional<String> cluster = Optional.ofNullable(dimensionsMap.get(CombinedMetricData.CLUSTER_KEY));
+
+        if (!service.isPresent()) {
+            service = Optional.ofNullable(setRecord.getService());
+        }
+
+        if (!cluster.isPresent()) {
+            cluster = Optional.ofNullable(setRecord.getCluster());
+        }
+
+        dimensionBuilder.put(CombinedMetricData.HOST_KEY, host.orElse(""));
+        dimensionBuilder.put(CombinedMetricData.SERVICE_KEY, service.orElse(""));
+        dimensionBuilder.put(CombinedMetricData.CLUSTER_KEY, cluster.orElse(""));
+
+        final ImmutableMap<String, String> dimensions = dimensionBuilder.build();
+
+        for (final Map.Entry<Statistic, CombinedMetricData.StatisticValue> record
+                : combinedMetricData.getCalculatedValues().entrySet()) {
+            final AggregatedData aggregatedData = new AggregatedData.Builder()
+                    .setFQDSN(new FQDSN.Builder()
+                            .setCluster(setRecord.getCluster())
+                            .setMetric(setRecord.getMetric())
+                            .setService(setRecord.getService())
+                            .setStatistic(record.getKey())
+                            .build())
+                    .setHost(host.get())
+                    .setIsSpecified(record.getValue().getUserSpecified())
+                    .setPeriod(combinedMetricData.getPeriod())
+                    .setPopulationSize(1L)
+                    .setSamples(Collections.emptyList())
+                    .setStart(combinedMetricData.getPeriodStart())
+                    .setSupportingData(record.getValue().getValue().getData())
+                    .setValue(record.getValue().getValue().getValue())
+                    .build();
+            builder.add(aggregatedData);
+        }
+        return Optional.of(new PeriodicData.Builder()
+                .setData(builder.build())
+                .setConditions(ImmutableList.of())
+                .setDimensions(dimensions)
+                .setPeriod(combinedMetricData.getPeriod())
+                .setStart(combinedMetricData.getPeriodStart())
+                .build());
+    }
+
+    private final Materializer _materializer;
+    private final Sink<AggregationMessage, CompletionStage<Done>> _sink;
+    private final Graph<FlowShape<HttpRequest, AggregationMessage>, NotUsed> _processGraph;
+
+    private static final Logger BAD_REQUEST_LOGGER =
+                LoggerFactory.getRateLimitLogger(HttpSourceActor.class, Duration.ofSeconds(30));
+    private static final Logger LOGGER = LoggerFactory.getLogger(HttpSourceActor.class);
+
+
+    private static class NoRecordsException extends RuntimeException {
```

Comment:
I don't think this should be a `RuntimeException`; I imagine it's this way to allow it to proceed through the flow?

Reply:
Nope, we can change this to a checked exception.  Changed it to inherit from IOException (as well as created another for invalid records).